### PR TITLE
HPCC-16541 jsocket / MP link closed tracing

### PR DESCRIPTION
**Pull request by jakesmith**
Add some tracing, inc. stack tracing when sockets lost
Also add an experimental option to reconenct the channel in some
cases.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln - please review

File: system/jlib/jsocket.cpp [hunk 1]
```diff
@@ -673,6 +665,22 @@ typedef union {
...
 #endif
 
+inline void LogErr(unsigned err,unsigned ref,const char *info,unsigned lineno,const char *tracename)
+{
+    if (err)
+    {
+        PROGLOG("jsocket(%d,%d)%s%s err = %d%s%s",ref,lineno,
+           (info&&*info)?" ":"",(info&&*info)?info:"",err,
+           (tracename&&*tracename)?" : ":"",(tracename&&*tracename)?tracename:"");
+        if (JSE_NOTCONN == err)
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Maybe also add JSE_CONNRESET and JSE_CONNABORTED here ?

File: thorlcr/slave/slavmain.cpp [hunk 2]
```diff
@@ -167,7 +167,10 @@ class CJobListener : public CSimpleInterface
         for (unsigned sc=1; sc<channelsPerSlave; sc++)
         {
             unsigned port = getMachinePortBase() + (sc * localThorPortInc);
-            mpServers.append(*startNewMPServer(port));
+            IMPServer *mpServer = startNewMPServer(port);
+            if (globals->getPropBool("@MPChannelReconnect"))
```
Comment (mckellyln) [inline thorlcr/slave/slavmain.cpp]: Could we move this line out of the loop and do it once ?

File: system/jlib/jsocket.cpp (hunk 1 above)
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: Could do, this issue has always been seen as a ENOTCONN though afaik.

File: system/mp/mpcomm.cpp [hunk 3]
```diff
@@ -1120,8 +1153,10 @@ protected: friend class CMPPacketReader;
                 try {
                     s->shutdown();
                 }
-                catch (IException *) { 
+                catch (IException *e) {
                     socketfailed = true; // ignore if the socket has been closed
+                    WARNLOG("closeSocket() : Ignoring shutdown error");
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp]: For now.
We I don't think we'll want to keep all the changes in this PR once/if we get to the bottom of this issue.

File: thorlcr/slave/slavmain.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/slave/slavmain.cpp]: could do, but it's very non time critical code here.

**Comment by mckellyln** [comment]
@jakesmith a few comments.
@richardkchapman ok to merge as is.

**Comment by jakesmith** [comment]
@mckellyln - added JSE_CONNRESET and JSE_CONNABORTED and move boolean out of channel initialization loop.

File: system/mp/mpcomm.cpp (hunk 3 above)
Comment (mckellyln) [inline system/mp/mpcomm.cpp]: Perhaps put this in a #define _TRACELINKCLOSED so its associated with this tracing effort.

**Comment by jakesmith** [comment]
@richardkchapman - ready to merge.

**Pull request merged**